doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+    This is analogous to "version vacuuming" in multiversioned databases.
+    As the system operates, the `since` frontier may advance to allow compaction of data.
+    This happens subject to constraints, and only when correctness permits.
+
+    We think of `since` as the "read frontier": times greater or equal to it may still be correctly read from the TVC.
+
+*   The `upper` frontier, which communicates that for times `time` not greater or equal to `upper`, the updates at `time` are complete.
+
+    The upper frontier provides the information about how a TVC may continue to evolve.
+    All future updates will necessarily have their `time` greater or equal to the `upper` frontier.
+    As the system operates, we expect the `upper` frontiers to advance indicating that the contents of the collection are known for more times.
+    This happens only when we become certain that updates for a time are complete.
+
+    We think of `upper` as the "write frontier": times greater or equal to it may still be written to the TVC.
+
+These two frontiers mean that for any `time` greater or equal to `since`, but not greater or equal to `upper`, we can exactly reconstruct the collection at `time`.
+If `time` is not greater or equal to `since`, we may have permanently lost historical distinctions that allow us to know the contents at `time`.
+If `time` is greater or equal to `upper`, we may still receive changes to the collection that occur at `time`.
+The two frontiers never move backward, and often move forward as the system operates.
+
+At each moment, Materialize's representation of a TVC is as a set of update triples `(data, time, diff)` and the two frontiers `since` and `upper`.
+We call this pair of set of updates and `(since, upper)` frontiers a *partial TVC* or pTVC.
+
+## Evolution
+
+Materialize maintains a (partial) map from `GlobalId`s to pTVCs.
+A `GlobalId` is never re-used to mean a different TVC.
+
+The life cycle of each `GlobalId` follows several steps:
+
+0. The `GlobalId` is initially not yet bound.
+
+    In this state, the `GlobalId` cannot be used, as we do not yet know what its initial `since` will be.
+
+1. The `GlobalId` is bound to some definition, with an initially equal `since` and `upper`.
+
+    This means that the collection is initially not useful, as it cannot be queried at any time greater or equal to `since` but not greater or equal to `upper`.
+    However, the initial value of `since` indicates that one should expect the data to eventually become available for each time greater or equal to `since`.
+    The partial TVC can now be used in queries, though results may not be immediately available.
+
+2. Repeatedly, either of the these two steps occur:
+
+    a. The `upper` frontier advances, revealing more of the future of the pTVC.
+
+    b. The `since` frontier advances, concealing some of the past of the pTVC.
+
+3. Eventually, the `since` frontier advances to the empty collection, rendering the TVC complete and closed.
```

Comment:
I am a bit suspicious that advancing to an empty collection should be sufficient for closing the TVC.
There might be scenarios where the contents of a source-backing table are re-populated periodically. For those cases, we will need to associate multiple TVCs with the same source.

Reply:
I'm sorry I don't understand. Can you explain more?

Reply:
I am thinking about the following scenario:

1. I create a table in postgres called `X`. 
1. At time `t_1`, I insert records `x(1), ..., x(n)`.
2. At time `t_2`, I delete records `x(1), ..., x(n)`.
1. At time `t_1`, I insert records `x(n+1), ..., x(n+m)`.

I then setup a Materialize source to observe changes in this table. I wonder whether Materialize will treat this sequence of actions as (a) one or more `GlobalId`s and (b) one or more TVCs.

Reply:
Oh ... I see the bug. I meant to write "the `since` frontier advances to the empty *frontier*", but I guess my brain said "don't repeat yourself" and I wrote "collection" instead of "set". Does it make sense with that edit?

Reply:
Meaning, when the *frontier* is the empty set we are done. The collection could be whatever, empty or non-empty. But when the `since` frontier is the empty set there are no times greater or equal to it, so it is impossible to use correctly any more.

Reply:
Ah, I see! Yes, in that case *empty frontier* definitely makes it clearer (at least if you assume some background knowledge of the lattice-backed structure of the `Timestamp` domain - you might want to give some extra hint otherwise).

Reply:
Changed to "... advances to the empty set of times" and changed the end to "rendering the TVC unreadable".